torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -314,6 +314,322 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+  }
+
+  void handle(ReductionOp* reduction_expr) final {
+    new ReductionOp(
+        reduction_expr->getReductionOpType(),
+        reduction_expr->init(),
+        consumer_,
+        reduction_expr->in());
+  }
+
+  void handle(BroadcastOp* broadcast_expr) final {
+    new BroadcastOp(consumer_, broadcast_expr->in());
+  }
+
+ private:
+  TensorView* consumer_;
+};
+
+// Create New Expr given producer - [an input for the expression]
+struct CreateExprProducer : public OptInDispatch {
+ public:
+  static void create(Expr* expr, TensorView* current, TensorView* producer) {
+    CreateExprProducer cep(current, producer);
+    cep.handle(expr);
+  }
+
+ private:
+  explicit CreateExprProducer(TensorView* current, TensorView* producer)
+      : current_(current), producer_(producer) {}
+
+  void handle(Expr* expr) final {
+    OptInDispatch::handle(expr);
+  }
+
+  void handle(UnaryOp* unary_expr) final {
+    new UnaryOp(unary_expr->getUnaryOpType(), unary_expr->out(), producer_);
+  }
+
+  void handle(BinaryOp* binary_expr) final {
+    if (binary_expr->lhs()->sameAs(current_)) {
+      new BinaryOp(
+          binary_expr->getBinaryOpType(),
+          binary_expr->out(),
+          producer_,
+          binary_expr->rhs());
+    } else {
+      new BinaryOp(
+          binary_expr->getBinaryOpType(),
+          binary_expr->out(),
+          binary_expr->lhs(),
+          producer_);
+    }
+  }
+
+  void handle(TernaryOp* ternary_expr) final {
+    if (ternary_expr->in1()->sameAs(current_)) {
+      new TernaryOp(
+          ternary_expr->getTernaryOpType(),
+          ternary_expr->out(),
+          producer_,
+          ternary_expr->in2(),
+          ternary_expr->in3());
+    } else if (ternary_expr->in2()->sameAs(current_)) {
+      new TernaryOp(
+          ternary_expr->getTernaryOpType(),
+          ternary_expr->out(),
+          ternary_expr->in1(),
+          producer_,
+          ternary_expr->in3());
+    } else {
+      new TernaryOp(
+          ternary_expr->getTernaryOpType(),
+          ternary_expr->out(),
+          ternary_expr->in1(),
+          ternary_expr->in2(),
+          producer_);
+    }
+  }
+
+  void handle(ReductionOp* reduction_expr) final {
+    new ReductionOp(
+        reduction_expr->getReductionOpType(),
+        reduction_expr->init(),
+        reduction_expr->out(),
+        producer_);
+  }
+
+  void handle(BroadcastOp* broadcast_expr) final {
+    new BroadcastOp(broadcast_expr->out(), producer_);
+  }
+
+ private:
+  TensorView* current_;
+  TensorView* producer_;
+};
+
+// Create New Expr given consumer - [output of the expression]
+void TensorView::createExprConsumer(Expr* expr, TensorView* consumer) {
+  CreateExprConsumer::create(expr, consumer);
+}
+
+// Create New Expr given producer - [an input for the expression]
+void TensorView::createExprProducer(
+    Expr* expr,
+    TensorView* current,
+    TensorView* producer) {
+  CreateExprProducer::create(expr, current, producer);
+}
```

Comment:
Can you please add more detailed comments about these two functions?

Reply:
```
  // In Cache Before, for the origin expr of the original tensor,
  // we create a new operation where the original tensor is replaced 
  // with the new cache tensor. This function creates a new expr
  // given the consumer, the output of the expression.
  void createExprConsumer(Expr* expr, TensorView* consumer);

  // In Cache After, for all the uses of the original tensor, we create
  // a new operation where the original tensor is replaced with the new
  // cache tensor. This function creates a new expr given a producer, 
  // an input for the expression.
  void createExprProducer(Expr* expr, TensorView* current, TensorView* producer);
```